src/docker/engine.rs

```diff
@@ -92,21 +98,98 @@ impl Engine {
 
 // determine if the container engine is docker. this fixes issues with
 // any aliases (#530), and doesn't fail if an executable suffix exists.
-fn get_engine_type(ce: &Path, msg_info: &mut MessageInfo) -> Result<EngineType> {
-    let stdout = Command::new(ce)
+fn get_engine_info(
+    ce: &Path,
+    msg_info: &mut MessageInfo,
+) -> Result<(EngineType, Option<Architecture>, Option<ContainerOs>)> {
+    let stdout_help = Command::new(ce)
         .arg("--help")
         .run_and_get_stdout(msg_info)?
         .to_lowercase();
 
-    if stdout.contains("podman-remote") {
-        Ok(EngineType::PodmanRemote)
-    } else if stdout.contains("podman") {
-        Ok(EngineType::Podman)
-    } else if stdout.contains("docker") && !stdout.contains("emulate") {
-        Ok(EngineType::Docker)
+    let kind = if stdout_help.contains("podman-remote") {
+        EngineType::PodmanRemote
+    } else if stdout_help.contains("podman") {
+        EngineType::Podman
+    } else if stdout_help.contains("docker") && !stdout_help.contains("emulate") {
+        EngineType::Docker
     } else {
-        Ok(EngineType::Other)
-    }
+        EngineType::Other
+    };
+
+    let mut cmd = Command::new(ce);
+    cmd.args(&["version", "-f \"{{ .Server.Os }},,,{{ .Server.Arch }}\""]);
+
+    let out = cmd.run_and_get_output(msg_info)?;
+
+    let stdout = out.stdout()?.to_lowercase();
+
+    let osarch = stdout
+        .split_once(",,,")
+        .map(|(os, arch)| -> Result<_> { Ok((ContainerOs::new(os)?, Architecture::new(arch)?)) })
+        .transpose();
+
+    let osarch = match (kind, osarch) {
+        (_, Ok(Some(osarch))) => Some(osarch),
+        (EngineType::PodmanRemote | EngineType::Podman, Ok(None)) => get_podman_info(ce, msg_info)?,
+        (_, Err(e)) => {
+            return Err(e.wrap_err(format!(
+                "command `{}` returned unexpected data",
+                cmd.command_pretty(msg_info, |_| false)
+            )));
+        }
+        (EngineType::Docker | EngineType::Other, Ok(None)) => None,
+    };
+
+    let osarch = if osarch.is_some() {
+        osarch
+    } else if !out.status.success() {
+        get_custom_info(ce, msg_info).with_error(|| {
+            cmd.status_result(msg_info, out.status, Some(&out))
+                .expect_err("status_result should error")
+        })?
+    } else {
+        get_custom_info(ce, msg_info)?
+    };
+
+    let (os, arch) = osarch.map_or(<_>::default(), |(os, arch)| (Some(os), Some(arch)));
+    Ok((kind, arch, os))
+}
+
+fn get_podman_info(
+    ce: &Path,
+    msg_info: &mut MessageInfo,
+) -> Result<Option<(ContainerOs, Architecture)>> {
+    let mut cmd = Command::new(ce);
+    cmd.args(&["info", "-f \"{{ .Version.OsArch }}\""]);
+    cmd.run_and_get_stdout(msg_info)
+        .map(|s| {
+            s.to_lowercase()
+                .split_once('/')
```

Comment:
This is being parsed incorrectly on WSL2: it's giving me:

```bash
+ echo '# Cross.toml
[build]
default-target = "x86_64-unknown-linux-musl"

[target."x86_64-unknown-linux-musl"]
image.name = "alpine:edge"
image.toolchain = ["x86_64-unknown-linux-musl"]
pre-build = ["apk add --no-cache gcc musl-dev"]'
+ /home/ahusz/git/cross/target/debug/cross run -v
+ cargo metadata --format-version 1
+ rustc --print sysroot
+ /usr/bin/podman
+ /usr/bin/podman version '-f "{{ .Server.Os }},,,{{ .Server.Arch }}"'
+ /usr/bin/podman info '-f "{{ .Version.OsArch }}"'
Error:
   0: architecture  "linux is not supported
   1: unknown variant ` "linux`, expected one of `android`, `darwin`, `freebsd`, `illumos`, `linux`, `netbsd`, `solaris`, `windows`
```

Reply:
We shouldn't need the `"-f \"{{ ... }\""` since they're already passed to the command without shell substitution. The above fix worked:

```diff
diff --git a/src/docker/engine.rs b/src/docker/engine.rs
index 74f4284..9bc37c4 100644
--- a/src/docker/engine.rs
+++ b/src/docker/engine.rs
@@ -118,13 +118,14 @@ fn get_engine_info(
     };
 
     let mut cmd = Command::new(ce);
-    cmd.args(&["version", "-f \"{{ .Server.Os }},,,{{ .Server.Arch }}\""]);
+    cmd.args(&["version", "-f", "{{ .Server.Os }},,,{{ .Server.Arch }}"]);
 
     let out = cmd.run_and_get_output(msg_info)?;
 
     let stdout = out.stdout()?.to_lowercase();
 
     let osarch = stdout
+        .trim()
         .split_once(",,,")
         .map(|(os, arch)| -> Result<_> { Ok((ContainerOs::new(os)?, Architecture::new(arch)?)) })
         .transpose();
@@ -161,10 +162,11 @@ fn get_podman_info(
     msg_info: &mut MessageInfo,
 ) -> Result<Option<(ContainerOs, Architecture)>> {
     let mut cmd = Command::new(ce);
-    cmd.args(&["info", "-f \"{{ .Version.OsArch }}\""]);
+    cmd.args(&["info", "-f", "{{ .Version.OsArch }}"]);
     cmd.run_and_get_stdout(msg_info)
         .map(|s| {
             s.to_lowercase()
+                .trim()
                 .split_once('/')
                 .map(|(os, arch)| -> Result<_> {
                     Ok((ContainerOs::new(os)?, Architecture::new(arch)?))
@@ -179,10 +181,11 @@ fn get_custom_info(
     msg_info: &mut MessageInfo,
 ) -> Result<Option<(ContainerOs, Architecture)>> {
     let mut cmd = Command::new(ce);
-    cmd.args(&["info", "-f \"{{ .Client.Os }},,,{{ .Client.Arch }}\""]);
+    cmd.args(&["info", "-f", "{{ .Client.Os }},,,{{ .Client.Arch }}"]);
     cmd.run_and_get_stdout(msg_info)
         .map(|s| {
             s.to_lowercase()
+                .trim()
                 .split_once(",,,")
                 .map(|(os, arch)| -> Result<_> {
                     Ok((ContainerOs::new(os)?, Architecture::new(arch)?))

```